src/plugin.rs

```diff
@@ -0,0 +1,63 @@
+use libloading::Library;
+use once_cell::sync::OnceCell;
+use std::ffi::{CString, OsStr};
+use std::ptr;
+use tree_sitter::Node;
+
+#[cfg(unix)]
+type RawSymbol<TExportedSymbol> = libloading::os::unix::Symbol<TExportedSymbol>;
+#[cfg(windows)]
+type RawSymbol<TExportedSymbol> = libloading::os::windows::Symbol<TExportedSymbol>;
+
+pub struct Filterer {
+    filterer: RawSymbol<unsafe extern "C" fn(*const Node) -> bool>,
+    _library: Library,
+}
+
+impl Filterer {
+    pub fn call(&self, node: &Node) -> bool {
+        unsafe { (self.filterer)(node) }
+    }
+}
+
+fn load_plugin(library_path: impl AsRef<OsStr>, filter_arg: Option<&str>) -> Filterer {
+    let library =
+        unsafe { Library::new(library_path).expect("Couldn't load filter dynamic library") };
+
+    if let Some(initialize) =
+        unsafe { library.get::<unsafe extern "C" fn(*const libc::c_char)>(b"initialize\0") }.ok()
+    {
+        let filter_arg = filter_arg.map(|filter_arg| {
+            CString::new(filter_arg).expect("Couldn't convert provided filter arg to CString")
+        });
+        unsafe {
+            initialize(
+                filter_arg
+                    .as_ref()
+                    .map_or_else(|| ptr::null(), |filter_arg| filter_arg.as_ptr()),
```

Comment:
> I don't understand why this isn't equivalent to the below but when I tried this it was passing an empty (C) string to the plugin?

Ok as discussed the issue with how I had this was presumably that the `.map_or_else()` was consuming `filter_arg` and thus the `.as_ptr()` was returning a dangling pointer. So in theory then doing `filter_arg.as_ref().map_or_else(...)` should work (since `filter_arg` is still alive, so the pointer into it should still be valid), and I tested it out and does appear to work correctly